File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java

class Solution {
    public int lengthOfLongestSubstring(String s) {
        // In Java, we use HashSet for O(1) average lookup
        Set<Character> st = new HashSet<>();
        int l = 0, maxi = 0, n = s.length();

        for (int i = 0; i < n; i++) {
            char currentChar = s.charAt(i);

            // Equivalent to while(st.count(s[i]))
            while (st.contains(currentChar)) {
                st.remove(s.charAt(l));
                l++;
            }
            
            st.add(currentChar);
            maxi = Math.max(maxi, (i - l + 1));
        }
        
        return maxi;
    }
}

File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        int maxi=0;
        int j=0;
        int n=s.Length;
        Dictionary<char,int>freq=new();
        for(int i=0;i<n;i++){
            while(freq.ContainsKey(s[i])){
                freq.Remove(s[j]);
                j++;
            }
            freq[s[i]]=1;
            maxi=Math.Max(maxi,i-j+1);   
        }
        return maxi;
    }
}